python/python/bystro/ancestry/ancestry_types.py

```diff
@@ -0,0 +1,158 @@
+"""Classes for common shapes of data in ancestry."""
+
+import attr
+from attr import field
+from attr.validators import ge, instance_of, le
+
+# Attr classes can throw either TypeError or ValueError upon receipt
+# of bad data.  Generally we won't care which, so define the union of
+# the two exceptions to catch all validation errors.  Because we'll be
+# using it in try blocks, this wants to be a tuple rather than a union.
+
+AttrValidationError = (ValueError, TypeError)
+
+
+#  It seems a bit overkill to have an attr class just for vcfs, but
+#  the validator needs to be shared between classes. So it just lives
+#  here as a pure python function with a weird type signature in order
+#  to support what attr expects from its validators.
+
+
+def _vcf_validator(_self: object, _attribute: attr.Attribute, value: str) -> None:
+    if not isinstance(value, str):
+        err_msg = f"vcf_path must be of type str, got: {value} instead"
+        raise TypeError(err_msg)
+    if not value.endswith(".vcf"):
```

Comment:
This will fail with compressed vcfs. We can create a pre-computed regex of
```
regex = re.compile( r'vcf\.{0,1}\w*')

def _vcf_validator(_self: object, _attribute: attr.Attribute, value: str) -> None:
    if not isinstance(value, str):
        err_msg = f"vcf_path must be of type str, got: {value} instead"
        raise TypeError(err_msg)
    if not regex.fullmatch(value):
   ...
```

Reply:
We shouldn't mind accepting a file that doesn't end in .vcf as long as it is a vcf file. But let's wait on relaxing that until we have a use case, the extension check is simple.